runtime/compiler/control/JITServerHelpers.cpp

```diff
@@ -44,84 +47,309 @@ TR::Monitor *JITServerHelpers::_clientStreamMonitor = NULL;
+      }
+   }
+
+// Copies original string into its location in the packed ROMClass and updates the SRP to it
+static void
+packCallback(const J9ROMClass *romClass, const J9SRP *origSrp, const char *slotName, ROMClassPackContext &ctx)
+   {
+   // Skip SRPs stored outside of the ROMClass bounds such as the ones in out-of-line method debug info
+   if (!ctx.isInline(origSrp, romClass))
+      return;
+
+   auto str = NNSRP_PTR_GET(origSrp, const J9UTF8 *);
+   auto srp = (J9SRP *)((uint8_t *)ctx.packedRomClass + ((uint8_t *)origSrp - (uint8_t *)romClass));
+
+   // Zero out skipped string SRPs
+   if (shouldSkipSlot(slotName))
+      {
+      TR_ASSERT(ctx.strToOffsetMap.find(str) != ctx.strToOffsetMap.end(),
+                "UTF8 slot %s not visited in 1st pass", slotName);
+      *srp = 0;
+      return;
+      }
+
+   auto it = ctx.strToOffsetMap.find(str);
+   TR_ASSERT(it != ctx.strToOffsetMap.end(), "UTF8 slot %s not visited in 1st pass", slotName);
+   auto dst = (uint8_t *)ctx.packedRomClass + (ctx.utf8SectionStart - (uint8_t *)romClass) + it->second;
+
+   NNSRP_PTR_SET(srp, dst);
+   if (dst == ctx.cursor)
+      ctx.cursor += copyUTF8((J9UTF8 *)dst, str);
+   else
+      TR_ASSERT((dst < ctx.cursor) && (memcmp(dst, str, J9UTF8_TOTAL_SIZE(str)) == 0), "Must be already copied");
+   }
+
+static void
+utf8SlotCallback(const J9ROMClass *romClass, const J9SRP *srp, const char *slotName, void *userData)
+   {
+   auto &ctx = *(ROMClassPackContext *)userData;
+   if (*srp)
+      ctx.callback(romClass, srp, slotName, ctx);
+   }
+
+// Invoked for each slot in a ROMClass. Calls ctx.callback() for all non-null SRPs to UTF8 strings.
+static void
+slotCallback(J9ROMClass *romClass, uint32_t slotType, void *slotPtr, const char *slotName, void *userData)
+   {
+   switch (slotType)
+      {
+      case J9ROM_UTF8:
+         utf8SlotCallback(romClass, (const J9SRP *)slotPtr, slotName, userData);
+         break;
+
+      case J9ROM_NAS:
+         if (auto nas = SRP_PTR_GET(slotPtr, const J9ROMNameAndSignature *))
+            {
+            utf8SlotCallback(romClass, &nas->name, slotName, userData);
+            utf8SlotCallback(romClass, &nas->signature, slotName, userData);
+            }
+         break;
+      }
+   }
+
+static bool
+isArrayROMClass(const J9ROMClass *romClass)
+   {
+   if (!J9ROMCLASS_IS_ARRAY(romClass))
+      return false;
+
+   auto name = J9ROMCLASS_CLASSNAME(romClass);
+   TR_ASSERT((name->length == 2) && (name->data[0] == '['),
+             "Unexpected array ROMClass name: %.*s", name->length, name->data);
+   return true;
+   }
+
+// Array ROMClasses have a different layout (see runtime/vm/romclasses.c):
+// they all share the same SRP array of interfaces, which breaks the generic
+// packing implementation. Instead, we manually pack all the data stored
+// outside of the ROMClass header: class name, superclass name, interfaces.
+// This function returns the total size of the packed array ROMClass.
```

Comment:
Does this mean that for array ROMClasses we may send interface names multiple times?

Reply:
For a given array ROMClass (or any ROMClass) we send all the strings it refers to (a single copy of each string). Overall, if we consider all the ROMClasses for a given client, there will be multiple copies of some strings (ones that were interned at the client). But it's the same for the current implementation - remote strings are retrieved and cached for each class separately. Deduplicating the strings on the server side would be non-trivial.